### PR TITLE
ui: allow list to be nested

### PR DESCRIPTION
**Pull request by meowcodes**
Screen Shot 2022-06-22 at 5 45 27 PM

- add with BoardItemAddButton and drag & drop
- '+ Board' button will be implemented after search update
- permission model will be tested after access is implemented
- release board full page view -> clicking on list title opens it

Screen Shot 2022-06-23 at 11 00 38 AM

### CODE REVIEW

File: querybook/webapp/components/BoardExpandableSection/BoardExpandableHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import { UrlContextMenu } from 'ui/ContextMenu/UrlContextMenu';
+import { ListLink } from 'ui/Link/ListLink';
+import { Title } from 'ui/Title/Title';
+
+interface IProps {
+    boardId: number;
+    boardName: string;
+    collapsed: boolean;
+    toggleCollapsed: () => void;
+    boardData?: IBoardWithItemIds;
```
Comment (czgu) [inline querybook/webapp/components/BoardExpandableSection/BoardExpandableHeader.tsx:21]: boardId, boardName can be inferred from boardData

File: querybook/webapp/components/BoardExpandableSection/BoardExpandableList.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+import {
+    IProcessedBoardItem,
+    BoardDraggableType,
+} from 'components/DataDocNavigator/navigatorConst';
+import { BoardListItemRow } from 'components/BoardExpandableSection/BoardListItemRow';
+
+import { DraggableList } from 'ui/DraggableList/DraggableList';
+import { IDragItem } from 'ui/DraggableList/types';
+
+export const BoardExpandableList: React.FunctionComponent<{
```
Comment (czgu) [inline querybook/webapp/components/BoardExpandableSection/BoardExpandableList.tsx:16]: interface IBoardExpandableListProps

File: querybook/webapp/components/BoardExpandableSection/BoardExpandableList.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    selectedDocId: number;
+    filterString: string;
+    boardId: number;
+    items: Array<{
+        boardItem: IBoardItem | { id: number };
```
Comment (czgu) [inline querybook/webapp/components/BoardExpandableSection/BoardExpandableList.tsx:21]: IBoardItem already covers { id: number }

File: querybook/webapp/components/BoardExpandableSection/BoardExpandableSection.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    const board = useSelector(
+        (state: IStoreState) => state.board.boardById[id]
+    );
+    const boardItemIds = board?.items;
+    const boardItemsSelector = React.useMemo(
+        () => makeBoardItemsSelector(),
+        []
+    );
+    const items = useSelector((state: IStoreState) =>
```
Comment (czgu) [inline querybook/webapp/components/BoardExpandableSection/BoardExpandableSection.tsx:58]: const items = useMakeSelector(makeBoardItemsSelector, board.id)

File: querybook/server/datasources/board.py [hunk 5]
```diff
@@ -155,7 +159,10 @@ def move_board_item(board_id, from_index, to_index):
     methods=["DELETE"],
 )
 def delete_board_item(board_id, item_type, item_id):
-    api_assert(item_type == "data_doc" or item_type == "table", "Invalid item type")
+    api_assert(
+        item_type == "data_doc" or item_type == "table" or item_type == "board",
```
[the remark this answers is not in the document]
Comment (meowcodes) [reply querybook/server/datasources/board.py:172]: this is gettting updated with the 'query' in a later pr

File: querybook/webapp/components/BoardExpandableSection/BoardExpandableHeader.tsx (hunk 1 above)
Comment (meowcodes) [reply querybook/webapp/components/BoardExpandableSection/BoardExpandableHeader.tsx:21]: not for public list

File: querybook/server/datasources/board.py [hunk 6]
```diff
@@ -135,6 +145,10 @@ def add_board_item(board_id, item_type, item_id):
             is None,
             "Item already exists",
         )
+        api_assert(
```
Comment (czgu) [inline querybook/server/datasources/board.py:164]: just realized this check can be done without any querying, can you put it after line 125

**Review by czgu — APPROVED** [review]

**Pull request merged**